js/viz.js

```diff
@@ -1,29 +1,111 @@
 // eslint-disable-next-line no-unused-vars
 const viz = {
-  draw(context, websites) {
-    const pi = Math.PI * 2;
-    let x = 50, y = 50;
-
-    context.fillStyle = 'white';
-    context.beginPath();
-
-    for (const website in websites) {
-      context.moveTo(x, y);
-      context.arc(x, y, 10, 0, pi);
-      context.fillText(website, x, y);
-
-      let x1 = x;
-      for (const thirdParty in websites[website].thirdPartyRequests) {
-        x1+=150;
-        context.moveTo(x1, y);
-        context.arc(x1, y, 5, 0, pi);
-        context.fillText(thirdParty, x1, y);
+  draw(nodesData, linksData) {
+    // set up area of the page to put the graph
+    const svg = d3.select('svg');
+    const width = svg.attr('width');
+    const height = svg.attr('height');
+
+    // set up simulation
+    const simulation = d3.forceSimulation().nodes(nodesData);
+
+    // add forces
+    simulation
+      // charge force treats each node as a charged particle,
+      // so that nodes will repel each other when they get too close together
+      .force('charge_force', d3.forceManyBody())
+      // centering force drives all nodes torward the center of the canvas
+      .force('center_force', d3.forceCenter(width/2, height/2))
+      // link force constrains the nodes' motion based on how they are linked
+      .force('links', d3.forceLink(linksData).id(function(d) {
+        return d.hostname;
+      }));
+
+    // separate first from third party nodes
+    const nodesFirstParty = [];
+    const nodesThirdParty = [];
+
+    for (let i = 0; i < nodesData.length; i++) {
+      if (nodesData[i].party === 'first') {
+        nodesFirstParty.push(nodesData[i]);
+      } else {
+        nodesData[i].type = d3.symbolTriangle;
+        nodesData[i].size = 100;
+        nodesThirdParty.push(nodesData[i]);
       }
+    }
+
+    // draw circles for the firstParty nodes
+    const nodeFirstParty = svg.append('g')
+      .attr('class', 'nodes')
+      .selectAll('circle')
+      .data(nodesFirstParty)
+      .enter()
+      .append('circle')
+      .attr('r', 5)
+      .attr('fill', 'white');
+
+    // draw triangles for the thirdParty nodes
+    const nodeThirdParty = svg.append('g')
+      .attr('class', 'nodes')
```

Comment:
> I need to use let instead of const and have several re-assignment statements following the declaration?

No not let, objects can be manipulated as I'm pretty sure I have seen you understand that. My understanding of your question is you are asking more about:

```
const nodeFirstParty = svg.append('g');
nodeFirstParty.attr('class', 'nodes')
const circles = nodeFirstParty.selectAll('circle')
circles.data(nodesFirstParty)
...
```

This is why I suggested this way, it's more explicit to someone not knowing d3 what the methods are operating on. (Also I'm guessing that is how the API works, it's not clear from the code at all as I don't know d3).